### ISSUE TITLE
Ray Serve performance 

### ISSUE THREAD

**Issue by satish-madugula** [ray-project/ray#20724]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Serve

### What happened + What you expected to happen

A keras/ Tensorflow model  deployed on ray serve, seemed to be working slow compared to general python inference code 

### Versions / Dependencies

Python: 3.8
Ray: 1.8
OS: Linux

### Reproduction script

Ray Serve is slower than a basic python script, I think the issue is with the "request type", Correct me if am wrong.
Deployed a model on ray serve, 

I am using a normal python script for requests as below, it takes around 4 minutes to complete the task for around 100 images. 

`for file in os.listdir(path):
    #image preprocessing
    resp = requests.get("http://localhost:9001/predict",json={"array":img.tolist()})
    prediction = np.array(resp.json()['prediction'])`

Also tried with  ray core for parallelising the preprocess and sending request but this too seemed slow

`@ray.remote()
def send_request(filepath):
    # image processing
    return requests.get("http://localhost:9001/predict", json={"array": img.tolist()})
resp = [send_request.remote(filepath) for filepath in os.listdir(path) ]
resp = ray.get(resp)
--check only for a single img prediction
resp = resp[0].json()
prediction = np.array(resp['prediction'])`

Using ray it took me around 4 mins to process 100 images, 

While using simple python script  (below)  took me around 50 seconds. 

`for file in os.listdir(path):
    ## image preprocessing
    prediction = predict_model.predict(img)
`

Is this because of the Http requests , if so how can I change it ?

### Anything else

Is there a possibility of using gRPC requests for ray serve ?

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by simon-mo** [comment]
how big are the predictions? the input looks small enough (just file path) but the output serialization and deserialization might be the bottleneck here

**Comment by jiaodong** [comment]
I would assume your final prediction is probably a classification result that's small, but compare with your simple python script that presumably directly reads the image file and call into prediction locally, with ray serve you're unpacking your image as python array and send it as json via http, which introduces significant additional work.

Just for side-by-side comparison, can you try to pass in the file path in your request to ray serve http endpoint where actual fetching happens in the deployment code ?

**Comment by satish-madugula** [comment]
@jiaodong, Thank you, sending file path as request did solve the problem (included the post processing of results in the deployment code). 
 what if I don't have access to the path to read the image ??

@simon-mo The output is a numpy array (1 x 31x63) 
Few things observed are (sending numpy array as request)

when num_cpus = 4 (For 100 images )
1. and the post processing of the results is done at the client side it is slow.
2. and the post processing of the results is done at the server side it is fast ( 58 secs)

**when num_cpus = 10, post processing of the results is done at the server side it took me around 2m 30 sec.**
**when num_cpus = 1, post processing of the results is done at the server side it took me around 22 sec.**

I'm not sure, if I'm doing any mistake, or is this normal ??

**Comment by jiaodong** [comment]
> what if I don't have access to the path to read the image ??

That's where I would take advantage of shared memory in ray cluster. You can start off by saving preprocessed images into ray cluster first via `obj_ref = ray.put(img)` then you can pass its binary representation `binary = obj_ref.binary()` around and reconstructed by `obj_ref = ray.ObjectRef(binary)` then you can fetch this remote object anywhere in the ray cluster by simply calling `img = ray.get([obj_ref])` again.

I would recommend checking out "Object resolution" section in our public whitepaper: https://docs.google.com/document/d/1lAy0Owi-vPz2jEqBSaHNQcy2IBSDEHyXNOQZlGuj93c/edit#heading=h.rtepjltxoeb7
